lib/Dialect/AIE/Transforms/AIEAssignLockIDs.cpp

```diff
@@ -40,50 +39,69 @@ struct AIEAssignLockIDsPass : AIEAssignLockIDsBase<AIEAssignLockIDsPass> {
     DeviceOp device = getOperation();
     OpBuilder rewriter = OpBuilder::atBlockEnd(device.getBody());
 
-    std::map<Operation *, std::pair<int, std::set<int>>> tileToLastID;
-
-    // The first pass scans for and stores the existing lockIDs. This data is
-    // stored in a map with the lock’s tile operation as the key, while the
-    // value to the key is a pair with the current potential lockID and a set
-    // that stores the currently assigned lockIDs.
-    for (auto lock : device.getOps<LockOp>()) {
-      if (lock.getLockID().has_value()) {
-        Operation *lock_tile = lock.getTile().getDefiningOp();
-        tileToLastID[lock_tile].first = 0;
-        tileToLastID[lock_tile].second.insert(lock.getLockIDValue());
-      }
-    }
+    // All of the lock ops on a tile, separated into ops which have been
+    // assigned to a lock, and ops which have not.
+    struct TileLockOps {
+      DenseSet<int> assigned;
+      SmallVector<LockOp> unassigned;
+    };
+
+    DenseMap<TileOp, TileLockOps> tileToLocks;
 
-    // The second pass scans for locks with no lockIDs and assigns locks.
-    for (auto lock : device.getOps<LockOp>()) {
-      Operation *lock_tile = lock.getTile().getDefiningOp();
-      if (!lock.getLockID().has_value()) {
-        if (tileToLastID.find(lock_tile) == tileToLastID.end()) {
-          // If the tile operation corresponding to the lock does not exist in
-          // the data structure, initialize the lockID with 0 with an empty set.
-          tileToLastID[lock_tile].first = 0;
-        } else if (tileToLastID[lock_tile].first < 15) {
-          // If the tile operation of the lock exists, the potential lockID is
-          // checked with the set containing occupied lockIDs until a lockID
-          // that is free is found.
-          int potential_ID = tileToLastID[lock_tile].first;
-          while (true) {
-            if (tileToLastID[lock_tile].second.find(potential_ID) !=
-                tileToLastID[lock_tile].second.end())
-              potential_ID++;
-            else
-              break;
+    // Construct data structure storing locks by tile.
+    for (auto lockOp : device.getOps<LockOp>()) {
+
+      auto tileOp = lockOp.getTileOp();
+      bool isAssigned = lockOp.getLockID().has_value();
+
+      if (isAssigned) {
+        auto lockID = lockOp.getLockID().value();
+        auto iter = tileToLocks.find(tileOp);
+        if (iter == tileToLocks.end()) {
+          tileToLocks.insert({tileOp, {{lockID}, /* unassigned = */ {}}});
+        } else {
+          if (iter->second.assigned.find(lockID) !=
+              iter->second.assigned.end()) {
+            auto diag = lockOp->emitOpError("is assigned to the same lock (")
+                        << lockID << ") as another op.";
+            diag.attachNote(tileOp.getLoc())
+                << "tile has lock ops assigned to same lock.";
+            return signalPassFailure();
```

Comment:
return diag;

Reply:
this is a `void runOnOperation()` - you can't return `diag` here.

Reply:
Ah, my mistake :)